go-controller/pkg/libovsdbops/portgroup.go

```diff
@@ -1,139 +1,55 @@
-	var aclUUIDs []string
 	if len(acls) > 0 {
-		aclUUIDs = make([]string, 0, len(acls))
+		pg.ACLs = make([]string, 0, len(acls))
 		for _, acl := range acls {
-			aclUUIDs = append(aclUUIDs, acl.UUID)
+			pg.ACLs = append(pg.ACLs, acl.UUID)
 		}
 	}
 
-	var portUUIDs []string
 	if len(ports) > 0 {
-		portUUIDs = make([]string, 0, len(ports))
+		pg.Ports = make([]string, 0, len(ports))
 		for _, port := range ports {
-			portUUIDs = append(portUUIDs, port.UUID)
-		}
-	}
-
-	return &nbdb.PortGroup{
-		Name:        hashName,
-		ACLs:        aclUUIDs,
-		Ports:       portUUIDs,
-		ExternalIDs: map[string]string{"name": name},
-	}
-}
-
-func createOrUpdatePortGroupOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, pg *nbdb.PortGroup) ([]libovsdb.Operation, error) {
-	if ops == nil {
-		ops = []libovsdb.Operation{}
-	}
-
-	err := findPortGroup(nbClient, pg)
-	if err != nil && err != libovsdbclient.ErrNotFound {
-		return nil, err
-	}
-
-	if err == libovsdbclient.ErrNotFound {
-		timeout := types.OVSDBWaitTimeout
-		condition := model.Condition{
-			Field:    &pg.Name,
-			Function: libovsdb.ConditionEqual,
-			Value:    pg.Name,
-		}
-		waitOps, err := nbClient.Where(pg, condition).Wait(libovsdb.WaitConditionNotEqual, &timeout, pg, &pg.Name)
-		if err != nil {
-			return nil, err
-		}
-		ops = append(ops, waitOps...)
-		op, err := nbClient.Create(pg)
-		if err != nil {
-			return nil, err
-		}
-		ops = append(ops, op...)
-		return ops, nil
-	}
-
-	mutations := []model.Mutation{}
-	if len(pg.Ports) > 0 {
-		mutations = append(mutations, model.Mutation{
-			Field:   &pg.Ports,
-			Mutator: libovsdb.MutateOperationInsert,
-			Value:   pg.Ports,
-		})
-	}
-	if len(pg.ACLs) > 0 {
-		mutations = append(mutations, model.Mutation{
-			Field:   &pg.ACLs,
-			Mutator: libovsdb.MutateOperationInsert,
-			Value:   pg.ACLs,
-		})
-	}
-	if len(pg.ExternalIDs) > 0 {
-		mutations = append(mutations, model.Mutation{
-			Field:   &pg.ExternalIDs,
-			Mutator: libovsdb.MutateOperationInsert,
-			Value:   pg.ExternalIDs,
-		})
-	}
-	if len(mutations) > 0 {
-		op, err := nbClient.Where(pg).Mutate(pg, mutations...)
-		if err != nil {
-			return nil, err
+			pg.Ports = append(pg.Ports, port.UUID)
 		}
-		ops = append(ops, op...)
-		return ops, nil
 	}
 
-	return ops, nil
+	return &pg
 }
 
+// CreateOrUpdatePortGroupsOps creates or updates the provided port groups
+// returning the corresponding ops
 func CreateOrUpdatePortGroupsOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, pgs ...*nbdb.PortGroup) ([]libovsdb.Operation, error) {
-	if ops == nil {
-		ops = []libovsdb.Operation{}
-	}
-	for _, pg := range pgs {
-		var err error
-		ops, err = createOrUpdatePortGroupOps(nbClient, ops, pg)
-		if err != nil {
-			return nil, err
+	opModels := make([]operationModel, 0, len(pgs))
+	for i := range pgs {
+		pg := pgs[i]
+		opModel := operationModel{
+			Model:            pg,
+			OnModelMutations: []interface{}{}, // update all
```

Comment:
This should be OnModelUpdates and not OnModelMutations. Nice catch.

Before we handled this as mutations out of precaution when this was the first libovsdb based implementation. Now there is clear semantics in place on what ClearOrUpdate does and other methods available to do mutations.

Now, this probably should have never worked, as it was, for an updated policy that would have added new ACLs  so need to go back and check what UT we have for this. 

Fixing as explained for now.

Reply:
Oh, makes sense. we have no use for this CreateOrUpdatePortGroupsOps as an update because when a policy is updated we handle it through destroy and create and not update. Otherwise the bug would have stood out